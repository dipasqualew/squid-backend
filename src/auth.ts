import jwt from 'jsonwebtoken';

export const SECRET = 'test';

export interface Claim {
  uuid: string;
  email: string;
  iat?: number;
}

export const isClaim = (obj: unknown): obj is Claim => {
  if (typeof obj !== 'object') {
    return false;
  }

  if (!obj) {
    return false;
  }

  return ('uuid' in obj) && ('email' in obj);
};

export const sign = (claim: Claim): Promise<string> => new Promise((resolve, reject) => {
  jwt.sign({ uuid: claim.uuid, email: claim.email }, SECRET, { algorithm: 'HS256' }, (err, token) => {
    if (err || !token) {
      return reject(err);
    }

    return resolve(token);
  });
});

export const verify = (token: string): Promise<Claim> => new Promise((resolve, reject) => {
  jwt.verify(token, SECRET, (err, decoded) => {
    if (err || !decoded) {
      return reject(err);
    }

    if (!isClaim(decoded)) {
      const error = new Error('Invalid JWT.');
      return reject(error);
    }

    return resolve(decoded);
  });
});
