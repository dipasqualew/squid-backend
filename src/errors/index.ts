export { BaseError } from './BaseError';
export { Http403 } from './http/Http403';
export { Http404 } from './http/Http404';
