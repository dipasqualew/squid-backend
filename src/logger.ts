

export class Logger {

  public log: (...unknown: unknown[]) => void;

  constructor(quiet = false) {
    this.log = quiet ? () => {} : console.log;
  }
}
