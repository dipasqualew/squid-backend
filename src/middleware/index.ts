export { AuthMiddleware } from './auth.middleware';
export { DatabaseMiddleware } from './db.middleware';
export { LoggerMiddleware } from './logger.middleware';

export { SquidErrorHandler } from './error.handler';
