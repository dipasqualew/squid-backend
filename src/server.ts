import type { Server } from 'http';

import Koa from 'koa';
import cors from '@koa/cors';
import bodyParser from 'koa-bodyparser';
import type knex from 'knex';

import { Logger } from './logger';
import { DatabaseMiddleware } from './db/config';
import { Claim, verify } from './auth';

// Routes
import { V1_USERS_AUTH_VIEWS } from './routes/users/auth.view';
import { V1_STATUS_VIEWS } from './routes/status.view';
import { SquidsController } from './routes/squids.views';

export interface AppConfig {
  logging?: {
    quiet: boolean,
  },
  server: {
    port: number;
  }
}

export type SquidAppState = Koa.DefaultState;

export interface SquidAppContext extends Koa.Context {
  user: Claim | null;
  db: knex;
}

export const AuthMiddleware = async (context: SquidAppContext, next: () => Promise<void>): Promise<void> => {
  const jwt = context.get('Authorization');

  if (jwt) {
    const decoded = await verify(jwt);
    context.user = decoded;
  } else {
    context.user = null;
  }

  await next();
};

export class App {
  private _app: Koa<SquidAppState, SquidAppContext> | null;
  public server: Server | null;
  public readonly config: AppConfig;
  public logger: Logger;

  constructor(config: AppConfig) {
    this.config = config;
    this._app = null;
    this.server = null;
    this.logger = new Logger(config.logging?.quiet ?? false);
  }

  get app(): Koa<SquidAppState, SquidAppContext> {
    if (!this._app) {
      this._app = new Koa<SquidAppState, SquidAppContext>();
      this._app.use(DatabaseMiddleware);
      this._app.use(AuthMiddleware);
      this._app.use(cors());
      this._app.use(bodyParser({ enableTypes: ['json'] }));
      this._app.use(V1_STATUS_VIEWS.routes());
      this._app.use(V1_USERS_AUTH_VIEWS.routes());
      this._app.use(SquidsController.routes());
    }

    return this._app;
  }

  listen(): Server {
    this.logger.log(`Listening on port: ${this.config.server.port}`);

    this.server = this.app.listen(this.config.server.port);

    return this.server;
  }
}

export const start = (port: number): () => Server => {
  const app = new App({ server: { port } });
  const server = app.listen();
  const close = () => server.close();

  return close;
};
