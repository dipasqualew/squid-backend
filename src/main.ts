import { App, AppConfig } from './server';

const APP_CONFIG: AppConfig = {
  server: {
    port: 3000,
  },
};

const app = new App(APP_CONFIG);

app.listen();
