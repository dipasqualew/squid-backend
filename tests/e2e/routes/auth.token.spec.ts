import { AxiosResponse } from 'axios';
import { v4 as uuid4 } from 'uuid';

import { App } from '../../../src/server';
import { getDB, query } from '../utils';
import { verify } from '../../../src/auth';

describe('e2e.routes: /auth/token', () => {
  const PORT = 30002;
  const app = new App({
    logging: { quiet: true },
    server: { port: PORT },
  });

  const user = {
    uuid: uuid4(),
    email: `${uuid4()}@dipasqualew.com`,
    password: uuid4(),
  };

  beforeAll(async () => {
    await getDB().table('users').insert([user]);
    await app.listen();
  });

  afterAll(async () => {
    await app.server?.close();
    await getDB().table('users').where('uuid', user.uuid).delete();
  });

  describe('With good credentials', () => {
    let response: AxiosResponse | null = null;
    const payload = {
      email: user.email,
      password: user.password,
    };

    beforeAll(async () => {
      response = await query('/v1/users/auth/jwt', PORT, 'POST', payload);
    });

    it('returns HTTP 200', () => {
      expect(response?.status).toEqual(200);
    });

    it('returns the JWT', async () => {
      const jwt = response?.data?.jwt as string;
      expect(jwt).toEqual(expect.any(String));

      const decoded = await verify(jwt);
      expect(decoded.uuid).toEqual(user.uuid);
      expect(decoded.email).toEqual(user.email);
      expect(decoded.iat).toEqual(expect.any(Number));
    });
  });

  describe('With bad credentials', () => {
    let response: AxiosResponse | null = null;
    const payload = {
      email: `${uuid4()}@dipasqualew.com`,
      password: uuid4(),
    };

    beforeAll(async () => {
      response = await query('/v1/users/auth/jwt', PORT, 'POST', payload);
    });

    it('returns HTTP 403', () => {
      expect(response?.status).toEqual(403);
    });

    it('returns an error list', () => {
      expect(response?.data?.errors).toMatchSnapshot();
    });
  });


});
