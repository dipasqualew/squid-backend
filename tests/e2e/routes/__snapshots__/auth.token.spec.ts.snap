// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e.routes: /auth/token With bad credentials returns an error list 1`] = `
Array [
  "Not Authorized",
]
`;
